src/analysis_and_optimization/Partial_evaluator.ml

```diff
@@ -677,9 +712,38 @@ let rec eval_expr (e : Middle.expr_typed_located) =
             Lit (Int, Int.to_string (Bool.to_int (r1 <> 0. || r2 <> 0.)))
         | e1', e2' -> EOr (e1', e2') )
       | Indexed (e, l) ->
-          (* TODO: do something clever with array and matrix expressions here?
-  Note  that we could also constant fold array sizes if we keep those around on declarations. *)
-          Indexed (eval_expr e, List.map ~f:eval_idx l) ) }
+          let rec eval_indexed_expr expr =
+            let rec eval_indexed_expr_one_step expr =
+              match expr with
+              | Indexed (e, l) -> (
+                match l with
+                | [] -> e.expr
```

Comment:
Same here - can you include this pattern into the top level pattern matching? Would be nice not to have pattern matches so deep if possible - a lot to hold in your head at once when reading. here,
```ocaml
              | Indexed (e, []) -> e.expr
              | Indexed ({expr=FunApp (t, f, elts); _}, [index])
                         when f = string_of_internal_fn FnMakeArray
                         || f = string_of_internal_fn FnMakeRowVec -> 
 ....

```